### ISSUE TITLE
PMML model has different prediction result with xgboost model

### ISSUE THREAD

**Issue by Ying456123** [jpmml/jpmml-xgboost#42]
@vruusmann 
I trained a xgboost model, and then using jpmml-xgboost-executable-1.3.11.jar and jpmml-sklearn-executable-1.5.18.jar to convert it to an pmml model. Then I found the two prediction results are not similar, could you please help to check? Thanks a lot.
train a xgboost model:
```
import numpy as np
from xgboost import XGBClassifier
from sklearn.model_selection import train_test_split
np.random.seed(0)
data = np.random.rand(15, 5)  # 5 entities, each contains 10 features
label = np.random.randint(2, size=15)  # binary target
X_train, X_test, y_train, y_test = train_test_split(data, label, random_state=0)
model = XGBClassifier(objective='binary:logistic',
                      learning_rate=0.2,
                      gamma=0.1,
                      n_estimators=200,
                      max_delta_step=3,
                      max_depth=8,
                      min_child_weight=2,
                      subsample=0.8,
                      colsample_bytree=0.8,
                      scale_pos_weight=1.2,
                      maximize=True,
                      random_state=27,
                      reg_alpha=0.0002,
                      reg_lambda=0.98,
                      base_score=0.6,
                      n_jobs=-1)
model = model.fit(X_train,
                  y_train,
                  eval_set=[(X_test, y_test)],
                  eval_metric="aucpr",
                  early_stopping_rounds=10)
print("prediction result:")
record = np.array([[0.5488135,0.71518937,0.60276338,0.54488318,0.4236548 ]])
print(model.predict(record))
print(model.predict_proba(record))
```
From the result, I see:
![image](https://user-images.githubusercontent.com/24699537/62354938-94c53f80-b540-11e9-8925-9eaa7f3d8279.png)

When I using the converted pmml model to do prediction, and I get this result:
```
import java.io.File
import com.spss.ac.acpmml.record.ScoreRecord
import com.spss.ac.acpmml.settings.ScoreSettings
import com.spss.ac.acpmml.util.ModelFactory

object PMMLTest {
  def main(args: Array[String]): Unit = {
    val xmlFile = new File("/Users/wangying/Downloads/SC_2019-08-01_16-44-42_GbtreeAucprExamplepmml/gbtreepmml.xml")
    //    val xmlFile = getXmlFile(pmmlFile)

    val settings = new ScoreSettings()
    val transformer = ModelFactory.createModel(xmlFile.getAbsolutePath, settings)
    val predictors = transformer.getPredictors
    val inputRecord = new ScoreRecord(predictors.size)
    val outputRecord = transformer.score(inputRecord, false)
    print(outputRecord)

  }
}
```
![image](https://user-images.githubusercontent.com/24699537/62355078-dbb33500-b540-11e9-92ae-300b20f03801.png)

thanks a lot.
also attached the xgboost model and converted pmml model
[models.tar.gz](https://github.com/jpmml/jpmml-xgboost/files/3460258/models.tar.gz)

**Comment by vruusmann** [comment]
@Ying456123 I can assure you that JPMML-XGBoost is generating correct and valid PMML documents. There is ample integration testing coverage included in the project, and related projects (eg. JPMML-SkLearn, JPMML-R).

Your toy example is broken - the model didn't learn anything useful. For example, you specify `n_estimators = 200`, but the exported PMML document contains only 11 member decision trees. Moreover, all these member decision trees make constant predictions - they don't use any independent variables (aka features).

Moreover, if you have problem with SPSS software, then you should be contacting SPSS support, not me.

**Comment by Ying456123** [comment]
@vruusmann Thanks a lot for your reply. My model indeed not learn that 200 estimators. But the question is not related with this.
I calculate with the pmml file manually, and find xgboost's new feature doesn't including in the PMML file.
The model I trained used parameter early_stopping_rounds. In xgboost-0.82, the model will use best_ntree_limit if it has this parameter (https://github.com/dmlc/xgboost/pull/3445). But the PMML file still uses all the eleven member to do the prediction, which causes the different final results.
```
using first tree: 1/(1+np.exp(0.374430223)) = 0.40747095882385276 (xgboost's result)
using 11 trees: 1/(1+np.exp(-0.0873479677)) = 0.5218231184276924 (PMML's result)
```
Could you please help to fix this? Or do you have other advice? Thanks a lot.

**Comment by Ying456123** [comment]
I find '--ntree-limit' in jpmml-xgboost, seems like jpmml-sklearn doesn't contain this option.

**Comment by vruusmann** [comment]
> I find '--ntree-limit' in jpmml-xgboost, seems like jpmml-sklearn doesn't contain this option.

This option is propagated to JPMML-SkLearn as `ntree_limit` conversion option:
https://github.com/jpmml/jpmml-sklearn/blob/master/src/main/java/xgboost/sklearn/BoosterUtil.java#L50

If you want to limit the number of trees, then you'd need to set this attribute manually. For example, when using the SkLearn2PMML package, then it's possible to set conversion options using the `PMMLPipeline.configure(**pmml_options)` method:
```python
pipeline = PMMLPipeline([
  ("estimator", XGBClassifier())
])
pipeline.fit(X, y)

# THIS
pipeline.configure(ntree_limit = 2)

sklearn2pmml(pipeline, "XGBoost.pmml")
```

If you're not using SkLearn2PMML package, then simply copy the business logic of this method to your Python script:
https://github.com/jpmml/sklearn2pmml/blob/master/sklearn2pmml/pipeline/__init__.py#L84-L88

**Comment by Ying456123** [comment]
@vruusmann Thanks a lot for your advice, but I think this parameter maybe should not be configured manually for sklearn api.  If I set 'ntree_limit' parameter in jpmml-xgboost is correct. Since for booster's predict function, tree_limit means:
```
ntree_limit (int) – Limit number of trees in the prediction; defaults to 0 (use all trees)
```
But for xgboost sklearn api's predict function, tree_limit means:
```
ntree_limit (int) – Limit number of trees in the prediction; defaults to best_ntree_limit if defined (i.e. it has been trained with early stopping), otherwise 0 (use all trees).
```
link: https://xgboost.readthedocs.io/en/latest/python/python_api.html#module-xgboost.sklearn
So for sklearn api, when it do prediction,  it will try to find best_ntree_limit even ntree_limit is None.
Below is the official code:
```
        if ntree_limit is None:
            ntree_limit = getattr(self, "best_ntree_limit", 0)
        return self.get_booster().predict(test_dmatrix,
                                          output_margin=output_margin,
                                          ntree_limit=ntree_limit,
                                          validate_features=validate_features)
```
https://github.com/dmlc/xgboost/blob/master/python-package/xgboost/sklearn.py#L462
This feature is added in xgboost-0.82 https://github.com/dmlc/xgboost/pull/3445
What's your opinion? thanks a lot for your patience.

**Comment by vruusmann** [comment]
> This feature is added in xgboost-0.82 dmlc/xgboost#3445

I haven't been following XGBoost library updates so closely, so I wasn't aware of the `best_ntree_limit` attribute.

**Comment by Ying456123** [comment]
Thanks a lot for your quick reply. Do you have a rough estimate about how long will this new feature issue can be closed? Thanks a lot

**Comment by vruusmann** [comment]
@Ying456123 Support for the `XGBModel.best_ntree_limit` attribute is included in SkLearn2PMML version 0.49.0+

**Comment by Ying456123** [comment]
@vruusmann Thanks a lot, I tried the jpmml-sklearn-executable-1.5.20.jar, now it's giving me the same result.
